CMakeConfigDeps and build type

Hi
Conan 2.15.0
OS: WSL2 Ubuntu 22.04

I would like to be able to install my consumer recipe dependencies for build type "release" and generate CMakeConfigDeps files for those.
But I would like to be able to use these with my consumer project building with another build type (e.g. "Debug").

This was no problem with the "old" CMakeDeps, but does not seem to work with CMakeConfigDeps.

I have tried with:
```
    deps = CMakeConfigDeps(self)
    deps.configuration = "Debug"
    deps.generate()
```
But it only seems to work if build type is not set in the settings, which they are required to be!

With CMakeDeps i could do this: `conan install ... -s:h build_type=Release -s:h &:build_type=Debug ...` effectively telling the system to install dependencies in Release mode, but generate cmake build files for building consumer in Debug mode.


My investigations show that CMakeDeps uses the build_type of the conanfile thus applying the consumer build type ("configuration" in the return statement):
```
class TargetConfigurationTemplate(CMakeDepsFileTemplate):
...
    @property
    def context(self):
...
        return {"pkg_name": self.pkg_name,
                "root_target_name": self.root_target_name,
                "config_suffix": self.config_suffix,
                "config": self.configuration.upper(),
                "deps_targets_names": ";".join(deps_targets_names),
                "components_names": components_names,
                "configuration": self.cmakedeps.configuration,
                "set_interface_link_directories": auto_link and is_win}
```
...whereas CMakeConfigDeps uses the build type of each individual dependency ("config" in the filename function):
```
class TargetConfigurationTemplate2:
    """
    FooTarget-release.cmake
    """
    def __init__(self, cmakedeps, conanfile, require):
        self._cmakedeps = cmakedeps
        self._conanfile = conanfile  # The dependency conanfile, not the consumer one
        self._require = require
...
    @property
    def filename(self):
        f = self._cmakedeps.get_cmake_filename(self._conanfile)
        # Fallback to consumer configuration if it doesn't have build_type
        config = self._conanfile.settings.get_safe("build_type", self._cmakedeps.configuration)
        config = (config or "none").lower()
        build = "Build" if self._conanfile.context == CONTEXT_BUILD else ""
        return f"{f}-Targets{build}-{config}.cmake"

```


Regards
Klaus

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

Thanks for the feedback.
I cannot make it work.
See my minimal demo project: https://github.com/klausholstjacobsen/conan-transitive-test/tree/cmakeconfigdeps-test

Checkout and build with `conan create --profile:build profile --profile:host profile --build missing .`
The poetry environment in the project will install conan 2.15.0 and cmake 3.28.1

The error I get:
```
Run Build Command(s): /home/khj/.cache/pypoetry/virtualenvs/conan-transitive-test-2X8sX9KB-py3.10/lib/python3.10/site-packages/cmake/data/bin/cmake -E env VERBOSE=1 /usr/bin/gmake -f Makefile -j16
/home/khj/.cache/pypoetry/virtualenvs/conan-transitive-test-2X8sX9KB-py3.10/lib/python3.10/site-packages/cmake/data/bin/cmake -S/home/khj/.conan2/p/b/consu1c33c3d605959/b -B/home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release --check-build-system CMakeFiles/Makefile.cmake 0
/home/khj/.cache/pypoetry/virtualenvs/conan-transitive-test-2X8sX9KB-py3.10/lib/python3.10/site-packages/cmake/data/bin/cmake -E cmake_progress_start /home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release/CMakeFiles /home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release//CMakeFiles/progress.marks
/usr/bin/gmake  -f CMakeFiles/Makefile2 all
gmake[1]: Entering directory '/home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release'
/usr/bin/gmake  -f CMakeFiles/Test.dir/build.make CMakeFiles/Test.dir/depend
gmake[2]: Entering directory '/home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release'
cd /home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release && /home/khj/.cache/pypoetry/virtualenvs/conan-transitive-test-2X8sX9KB-py3.10/lib/python3.10/site-packages/cmake/data/bin/cmake -E cmake_depends "Unix Makefiles" /home/khj/.conan2/p/b/consu1c33c3d605959/b /home/khj/.conan2/p/b/consu1c33c3d605959/b /home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release /home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release /home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release/CMakeFiles/Test.dir/DependInfo.cmake "--color="
gmake[2]: Leaving directory '/home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release'
/usr/bin/gmake  -f CMakeFiles/Test.dir/build.make CMakeFiles/Test.dir/build
gmake[2]: Entering directory '/home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release'
[ 50%] Building CXX object CMakeFiles/Test.dir/main.cpp.o
/usr/bin/c++  -isystem /home/khj/.conan2/p/b/libcuc8eb1637e5d3e/p/include -m64 -g -std=gnu++17 -MD -MT CMakeFiles/Test.dir/main.cpp.o -MF CMakeFiles/Test.dir/main.cpp.o.d -o CMakeFiles/Test.dir/main.cpp.o -c /home/khj/.conan2/p/b/consu1c33c3d605959/b/main.cpp
/home/khj/.conan2/p/b/consu1c33c3d605959/b/main.cpp:3:10: fatal error: gtest/gtest.h: No such file or directory
    3 | #include <gtest/gtest.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/Test.dir/build.make:79: CMakeFiles/Test.dir/main.cpp.o] Error 1
gmake[2]: Leaving directory '/home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release'
gmake[1]: *** [CMakeFiles/Makefile2:86: CMakeFiles/Test.dir/all] Error 2
gmake[1]: Leaving directory '/home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release'
gmake: *** [Makefile:94: all] Error 2


consumer/0.0.0: ERROR:
Package '4607b30100700e8252574bb14fac2875d42b712d' build failed
consumer/0.0.0: WARN: Build folder /home/khj/.conan2/p/b/consu1c33c3d605959/b/build/Release
ERROR: consumer/0.0.0: Error in build() method, line 38
        cmake.build()
        ConanException: Error 2 while executing
```
The error you see is with gtest. But if I remove gtest it will show an error with boost instead! So, I do not think it is related specifically to gtest!

Full build log attached: [build.log](https://github.com/user-attachments/files/19616681/build.log)

Regards
Klaus

--- a/conan/tools/cmake/cmakedeps2/target_configuration.py
+++ b/conan/tools/cmake/cmakedeps2/target_configuration.py
@@ -80,7 +80,7 @@ class TargetConfigurationTemplate2:
                         assert required_pkg == required_comp
                         comp = None
                         default_target = f"{dep.ref.name}::{dep.ref.name}"  # replace_requires
-                        link = pkg_type is not PackageType.SHARED 
+                        link = pkg_type is not PackageType.SHARED
                     else:
                         comp = required_comp
                         default_target = f"{required_pkg}::{required_comp}"
@@ -364,6 +364,8 @@ class TargetConfigurationTemplate2:
         set_property(TARGET {{lib}} APPEND PROPERTY IMPORTED_CONFIGURATIONS {{config}})
         set_target_properties({{lib}} PROPERTIES IMPORTED_LOCATION_{{config}}
                               "{{lib_info["location"]}}")
+        {% elif lib_info.get("type") == "INTERFACE" %}
+        set_property(TARGET {{lib}} APPEND PROPERTY IMPORTED_CONFIGURATIONS {{config}})
         {% endif %}
         {% if lib_info.get("link_location") %}
         set_target_properties({{lib}} PROPERTIES IMPORTED_IMPLIB_{{config}}

--- a/test/functional/toolchains/cmake/cmakedeps2/test_cmakeconfigdeps_new.py
+++ b/test/functional/toolchains/cmake/cmakedeps2/test_cmakeconfigdeps_new.py
@@ -1108,7 +1108,7 @@ class TestProtobuf:
         assert "protobuf: Release!" in c.out
 
 
-@pytest.mark.tool("cmake", "3.23")
+@pytest.mark.tool("cmake", "3.27")
 class TestConfigs:
     @pytest.mark.skipif(platform.system() != "Windows", reason="Only MSVC multi-conf")
     def test_multi_config(self, matrix_client):
@@ -1137,9 +1137,9 @@ class TestConfigs:
 
         # With modern CMake > 3.26 not necessary set(CMAKE_MAP_IMPORTED_CONFIG_DEBUG Release)
         cmake = textwrap.dedent("""
-            cmake_minimum_required(VERSION 3.15)
+            cmake_minimum_required(VERSION 3.27)
             project(app CXX)
-            set(CMAKE_MAP_IMPORTED_CONFIG_DEBUG Release)
+            # set(CMAKE_MAP_IMPORTED_CONFIG_DEBUG Release)
             find_package(matrix CONFIG REQUIRED)
 
             add_executable(app src/app.cpp src/main.cpp)
@@ -1155,6 +1155,41 @@ class TestConfigs:
         assert "matrix/1.0: Hello World Release!" in c.out
         assert "app/0.1: Hello World Debug!" in c.out
 
+    def test_cross_config_components(self, matrix_client_components):
+        # Release dependencies, but compiling app in Debug
+        c = matrix_client_components
+        c.run("new cmake_exe -d name=app -d version=0.1 -d requires=matrix/1.0")
+        app_cpp = textwrap.dedent("""
+            #include "app.h"
+            #include "module.h"
+            void app(){
+               module();
+            }""")
+        c.save({"src/app.cpp": app_cpp,
+                "src/main.cpp": gen_function_cpp(name="main", includes=["app"], calls=["app"])})
+
+        c.run(f"install . -s &:build_type=Debug -c tools.cmake.cmakedeps:new={new_value}")
+        # With modern CMake > 3.26 not necessary set(CMAKE_MAP_IMPORTED_CONFIG_DEBUG Release)
+        cmake = textwrap.dedent("""
+            cmake_minimum_required(VERSION 3.27)
+            project(app CXX)
+            # set(CMAKE_MAP_IMPORTED_CONFIG_DEBUG Release)
+            find_package(matrix CONFIG REQUIRED)
+
+            add_executable(app src/app.cpp src/main.cpp)
+            target_link_libraries(app PRIVATE matrix::matrix)
+            """)
+        c.save({"CMakeLists.txt": cmake})
+
+        preset = "conan-default" if platform.system() == "Windows" else "conan-debug"
+        c.run_command(f"cmake --preset {preset}")
+        c.run_command("cmake --build --preset conan-debug")
+
+        c.run_command(os.path.join("build", "Debug", "app"))
+        assert "main: Debug!" in c.out
+        assert "module: Release!" in c.out
+        assert "vector: Release!" in c.out
+
     @pytest.mark.skipif(platform.system() == "Windows", reason="This doesn't work in MSVC")
     def test_cross_config_implicit(self, matrix_client):
         # Release dependencies, but compiling app in Debug, without specifying it
@@ -1164,9 +1199,9 @@ class TestConfigs:
 
         # With modern CMake > 3.26 not necessary set(CMAKE_MAP_IMPORTED_CONFIG_DEBUG Release)
         cmake = textwrap.dedent("""
-            cmake_minimum_required(VERSION 3.15)
+            cmake_minimum_required(VERSION 3.27)
             project(app CXX)
-            set(CMAKE_MAP_IMPORTED_CONFIG_DEBUG Release)
+            # set(CMAKE_MAP_IMPORTED_CONFIG_DEBUG Release)
             find_package(matrix CONFIG REQUIRED)
 
             add_executable(app src/app.cpp src/main.cpp)

